src/Tools/TestDiscoveryWorker/Program.cs

```diff
@@ -0,0 +1,118 @@
+// Display the read text to the console
+string? output;
+
+// Wait for 'sync message' from the server.
+do
+{
+    output = sr.ReadLine();
+}
+while (!(output?.StartsWith("ASSEMBLY") == true));
+
+if ((output = sr.ReadLine()) is not null)
+{
+    var assemblyFileName = output;
+
+#if NET6_0_OR_GREATER   
+    var resolver = new System.Runtime.Loader.AssemblyDependencyResolver(assemblyFileName);
+    System.Runtime.Loader.AssemblyLoadContext.Default.Resolving += (context, assemblyName) =>
+    {
+        var assemblyPath = resolver.ResolveAssemblyToPath(assemblyName);
+        if (assemblyPath != null)
+        {
+            return Assembly.LoadFrom(assemblyPath);
+        }
+
+        return null;
+    };
+#endif
+
+
+    using var xunit = new XunitFrontController(AppDomainSupport.IfAvailable, assemblyFileName, shadowCopy: false);
+    var configuration = ConfigReader.Load(assemblyFileName);
+    var sink = new Sink();
+    xunit.Find(includeSourceInformation: false, messageSink: sink,
+                discoveryOptions: TestFrameworkOptions.ForDiscovery(configuration));
+
+    var builder = ImmutableArray.CreateBuilder<string>();
+    await foreach (var fullyQualifiedName in sink.GetTestCaseNamesAsync())
+    {
+        builder.Add(fullyQualifiedName);
+    }
+
+    var testsToWrite = builder.Distinct().ToArray();
+#if NET6_0_OR_GREATER
+    Console.WriteLine($"Discovered {testsToWrite.Length} tests in {Path.GetFileName(assemblyFileName)} (.NET Core)");
+#else
+    Console.WriteLine($"Discovered {testsToWrite.Length} tests in {Path.GetFileName(assemblyFileName)} (.NET Framework)");
+#endif
+
+    var directory = Path.GetDirectoryName(assemblyFileName);
+    using var fileStream = File.Create(Path.Combine(directory!, "testlist.json"));
+    JsonSerializer.Serialize(fileStream, testsToWrite);
+}
+
+internal class Sink : IMessageSink
+{
+    public Sink()
+    {
+        _channel = Channel.CreateUnbounded<string>();
+    }
+
+    private readonly Channel<string> _channel;
+
+    public async IAsyncEnumerable<string> GetTestCaseNamesAsync()
+    {
+        while (await _channel.Reader.WaitToReadAsync(default).ConfigureAwait(false))
+        {
+            while (_channel.Reader.TryRead(out var item))
+            {
+                yield return item;
+            }
+        }
+    }
+
+    public bool OnMessage(IMessageSinkMessage message)
+    {
+        if (message is ITestCaseDiscoveryMessage discoveryMessage)
+        {
+            OnTestDiscovered(discoveryMessage);
+        }
+
+        if (message is IDiscoveryCompleteMessage)
+        {
+            _channel.Writer.Complete();
+        }
+
+        return true;
+    }
+
+    private void OnTestDiscovered(ITestCaseDiscoveryMessage testCaseDiscovered)
+    {
+        var fullName = $"{testCaseDiscovered.TestCase.TestMethod.TestClass.Class.Name}.{testCaseDiscovered.TestCase.TestMethod.Method.Name}";
+        _ = _channel.Writer.TryWrite(fullName);
```

Comment:
Doesn't seem like we should ignore this. If it fails to write think we need to error in some way.

Reply:
The channel is unbounded so this will always succeed unless you run out of memory, or you closed the channel and no longer care about additional tests. Using the sync call also avoids async overhead.

So maybe just comment the reason why the result is ignored?

Reply:
Right but at least one of those cases is "we had a coding error" and right now that would lead to us silently skipping tests.